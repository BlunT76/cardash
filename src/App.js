/**
 * MinimalCarDashboard
 * https://github.com/BlunT76/cardash
 * philippe Pereira
 * huhmiel@free.fr
 */

import React, { PureComponent } from 'react';
import {
  StatusBar, StyleSheet, View, PermissionsAndroid, Platform,
} from 'react-native';

import { Provider } from 'react-redux';
import { createStore } from 'redux';
import KeepAwake from 'react-native-keep-awake';
// import devToolsEnhancer from 'remote-redux-devtools'; // usefull while in dev
import gpsDataReducer from './store/Reducers';

import { widthPercentageToDP, heightPercentageToDP } from './util/getDimensions';
import Speedometer from './components/Speedometer';
import AudioPlayer from './components/AudioPlayer';
import TimeDisplay from './components/TimeDisplay';
import MaxSpeed from './components/MaxSpeed';
import KilometerCounter from './components/KilometerCounter';
import TrackTitleDisplay from './components/TrackTitleDisplay';

// Set an initial global state directly: DEV
// const store = createStore(gpsDataReducer, devToolsEnhancer({ realtime: true }));
// Set an initial global state directly: RELEASE
const store = createStore(gpsDataReducer);

const styles = StyleSheet.create({
  globalContainer: {
    width: widthPercentageToDP('100%'),
    height: heightPercentageToDP('100%'),
    backgroundColor: '#272822',
  },
  speedContainer: {
    width: widthPercentageToDP('100%'),
    height: heightPercentageToDP('75%'),
    maxHeight: heightPercentageToDP('75%'),
    fontFamily: 'digital-7',
  },
  bottomContainer: {
    width: widthPercentageToDP('100%'),
    height: heightPercentageToDP('25%'),
    flexDirection: 'row',
    justifyContent: 'space-evenly',
    maxHeight: heightPercentageToDP('25%'),
    elevation: 5,
    backgroundColor: '#272822',
  },
});

export default class App extends PureComponent {
  constructor(props) {
    super(props);
    this.watchId = null;
    this.state = {
      allowSpeed: false,
      allowPlayer: false,
    };
  }

  componentDidMount() {
    this.GetAllPermissions();
  }

  GetAllPermissions = () => {
    if (Platform.OS === 'android') {
      PermissionsAndroid.requestMultiple([
        PermissionsAndroid.PERMISSIONS.ACCESS_FINE_LOCATION,
        PermissionsAndroid.PERMISSIONS.READ_EXTERNAL_STORAGE,
      ]).then((result) => {
        if (result['android.permission.ACCESS_FINE_LOCATION'] === 'granted') {
          this.setState({
            allowSpeed: true,
          });
        }
        if (result['android.permission.READ_EXTERNAL_STORAGE'] === 'granted') {
          this.setState({
            allowPlayer: true,
          });
        }
      });
    }
  }

  render() {
    const { allowSpeed, allowPlayer, allowMaxSpeed } = this.state;

    return (
      <Provider store={store}>
        <View style={styles.globalContainer}>

          <StatusBar backgroundColor="#272822" barStyle="light-content" />
          <KeepAwake />

          { allowSpeed && <KilometerCounter /> }
          { allowSpeed && <Speedometer allowMS={allowMaxSpeed} /> }
          { allowPlayer && <TrackTitleDisplay />}

          <View style={styles.bottomContainer}>

            { allowSpeed && <TimeDisplay /> }
            { allowPlayer && <AudioPlayer /> }
            { allowSpeed && <MaxSpeed /> }

          </View>

        </View>
      </Provider>
    );
  }
}
