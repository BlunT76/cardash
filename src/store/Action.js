// src/js/actions/index.js
export const ADD_GPS_DATA = 'ADD_GPS_DATA';
export const addGpsData = (payload) => {
  return { type: ADD_GPS_DATA, payload }
};

export const ADD_OLD_GPS_DATA = 'ADD_OLD_GPS_DATA';
export const addOldGpsData = (payload) => {
  return { type: ADD_OLD_GPS_DATA, payload }
};

export const ADD_TIME = 'ADD_TIME';
export const addTime = (payload) => {
  return { type: ADD_TIME, payload }
};

export const ADD_MAXSPEED = 'ADD_MAXSPEED';
export const addMaxSpeed = (payload) => {
  return { type: ADD_MAXSPEED, payload }
};
