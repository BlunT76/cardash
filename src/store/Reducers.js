import { ADD_GPS_DATA } from './Action';

const initialState = {
  lat: 0,
  lng: 0,
  oldlat: 0,
  oldlng: 0,
  maxspeed: 0,
  gpsTime: '',
};

const gpsDataReducer = (state = initialState, action) => {
  switch (action.type) {
    case ADD_GPS_DATA:
      return {
        ...state,
        lat: action.payload.latitude || 0,
        lng: action.payload.longitude || 0,
      };
    case 'ADD_OLD_GPS_DATA':
      return {
        ...state,
        oldlat: action.payload.latitude,
        oldlng: action.payload.longitude,
      };
    case 'ADD_TIME': {
      const hour = new Date(action.payload).toLocaleTimeString('fr-FR');
      return {
        ...state,
        gpsTime: hour.substring(0, hour.length - 3),
      };
    }
    case 'ADD_MAXSPEED':
      return {
        ...state,
        maxspeed: action.payload,
      };

    default:
      return state;
  }
};
export default gpsDataReducer;
