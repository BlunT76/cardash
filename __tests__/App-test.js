import 'react-native';
import React from 'react';
import renderer from 'react-test-renderer';
import App from '../src/App';
// import AudioPlayer from '../src/components/AudioPlayer';
import * as actions from '../src/store/Action';
import gpsDataReducer from '../src/store/Reducers';

jest.mock('react-native-sound', () => '');
jest.mock('react-native-keep-awake', () => 'KeepAwake');
jest.mock('react-native-get-music-files', () => '');

// Components snapshot tests
describe('Snapshots tests', () => {
  it('App renders correctly', () => {
    const tree = renderer.create(<App />).toJSON();
    expect(tree).toMatchSnapshot();
  });
  // need to mock store
  // it('AudioPlayer renders correctly', () => {
  //   const tree = renderer.create(<AudioPlayer />).toJSON();
  //   expect(tree).toMatchSnapshot();
  // });
});

// Redux Action tests
describe('actions tests', () => {
  it('should create an action to add a gps data', () => {
    const payload = 'test';
    const expectedAction = {
      type: actions.ADD_GPS_DATA,
      payload,
    };
    expect(actions.addGpsData(payload)).toEqual(expectedAction);
  });

  it('should create an action to add a old gps data', () => {
    const payload = 'test';
    const expectedAction = {
      type: actions.ADD_OLD_GPS_DATA,
      payload,
    };
    expect(actions.addOldGpsData(payload)).toEqual(expectedAction);
  });

  it('should create an action to add time', () => {
    const payload = 'test';
    const expectedAction = {
      type: actions.ADD_TIME,
      payload,
    };
    expect(actions.addTime(payload)).toEqual(expectedAction);
  });

  it('should create an action to add maxspeed', () => {
    const payload = 'test';
    const expectedAction = {
      type: actions.ADD_MAXSPEED,
      payload,
    };
    expect(actions.addMaxSpeed(payload)).toEqual(expectedAction);
  });
});

// Redux reducers tests
describe('reducers tests', () => {
  it('should return the initial state', () => {
    expect(gpsDataReducer(undefined, {})).toEqual(
      {
        lat: 0,
        lng: 0,
        oldlat: 0,
        oldlng: 0,
        speed: 0,
        maxspeed: 0,
        gpsTime: '',
        trackName: '',
      },
    );
  });

  it('should handle ADD_GPS_DATA', () => {
    const payload = { latitude: 0.44, longitude: 0.44 };
    const action = { type: 'ADD_GPS_DATA', payload };
    expect(gpsDataReducer(undefined, action)).toEqual(
      {
        lat: 0.44,
        lng: 0.44,
        oldlat: 0,
        oldlng: 0,
        speed: 0,
        maxspeed: 0,
        gpsTime: '',
        trackName: '',
      },
    );
  });

  it('should handle ADD_OLD_GPS_DATA', () => {
    const payload = { latitude: 0.44, longitude: 0.44 };
    const action = { type: 'ADD_OLD_GPS_DATA', payload };
    expect(gpsDataReducer(undefined, action)).toEqual(
      {
        lat: 0,
        lng: 0,
        oldlat: 0.44,
        oldlng: 0.44,
        speed: 0,
        maxspeed: 0,
        gpsTime: '',
        trackName: '',
      },
    );
  });

  // disabled while no locale detection
  // it('should handle ADD_TIME', () => {
  //   const payload = 1558450230;
  //   const action = { type: 'ADD_TIME', payload };
  //   expect(gpsDataReducer(undefined, action)).toEqual(
  //     {
  //       lat: 0,
  //       lng: 0,
  //       oldlat: 0,
  //       oldlng: 0,
  //       speed: 0,
  //       maxspeed: 0,
  //       gpsTime: '01:54',
  //     },
  //   );
  // });

  it('should handle ADD_MAXSPEED', () => {
    const payload = 50;
    const action = { type: 'ADD_MAXSPEED', payload };
    expect(gpsDataReducer(undefined, action)).toEqual(
      {
        lat: 0,
        lng: 0,
        oldlat: 0,
        oldlng: 0,
        speed: 0,
        maxspeed: 50,
        gpsTime: '',
        trackName: '',
      },
    );
  });
});
