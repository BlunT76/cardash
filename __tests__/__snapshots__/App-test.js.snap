// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshots tests App renders correctly 1`] = `
<View
  style={
    Object {
      "backgroundColor": "#272822",
      "height": 1334,
      "width": 750,
    }
  }
>
  <KeepAwake />
  <View
    style={
      Object {
        "backgroundColor": "#272822",
        "elevation": 5,
        "flexDirection": "row",
        "height": 333.5,
        "justifyContent": "space-evenly",
        "maxHeight": 333.5,
        "width": 750,
      }
    }
  />
</View>
`;
